File: 4.2-procesamiento-datos/prueba-2/version-2/javascript/worker.js
function createDataStructure(size) {
  const start = performance.now();
  const data = new Int32Array(size);
  for (let i = 0; i < size; i++) {
      data[i] = Math.floor(Math.random() * 1000);
  }
  const end = performance.now();

  const memory = data.byteLength / (1024 * 1024);
  return { data, time: end - start, memory };
}

function calculateSum(data) {
  const start = performance.now();
  let totalSum = data.reduce((acc, val) => acc + val, 0);
  const end = performance.now();

  return { result: totalSum, time: end - start };
}

function calculateMean(data, totalSum) {
  const start = performance.now();
  const mean = totalSum / data.length;
  const end = performance.now();

  return { result: mean, time: end - start };
}

function calculateStd(data, mean) {
  const start = performance.now();
  let sumSqDiff = 0;
  for (let i = 0; i < data.length; i++) {
      let diff = data[i] - mean;
      sumSqDiff += diff * diff;
  }
  const variance = sumSqDiff / data.length;
  const std = Math.sqrt(variance);
  const end = performance.now();

  return { result: std, time: end - start };
}

self.onmessage = function (event) {
  const size = event.data.size;
  if (!size) {
      self.postMessage({ error: "Size not provided" });
      return;
  }

  try {
      const metrics = {};
      const startTotal = performance.now();

      // Create
      const createResult = createDataStructure(size);
      const dataMemoryMB = createResult.memory;
      metrics.create = { time: createResult.time, memory: dataMemoryMB };
      const data = createResult.data;

      // Sum
      const sumResult = calculateSum(data);
      metrics.sum = { time: sumResult.time, memory: dataMemoryMB };

      // Mean
      const meanResult = calculateMean(data, sumResult.result);
      metrics.mean = { time: meanResult.time, memory: dataMemoryMB };

      // Std
      const stdResult = calculateStd(data, meanResult.result);
      metrics.std = { time: stdResult.time, memory: dataMemoryMB };

      // Total
      const totalTime = performance.now() - startTotal;
      const totalTimeExact = createResult.time + sumResult.time + meanResult.time + stdResult.time;
      metrics.total = {
          time: totalTime,
          memory: dataMemoryMB,
          exact_time: totalTimeExact};
      self.postMessage({ results: metrics })

  } catch (e) {
      self.postMessage({ error: e.message, stack: e.stack });
  }
};

File: 4.2-procesamiento-datos/prueba-2/version-2/python/worker.py
import time
import numpy as np
from pyscript import sync


def create_data_structure(size):
    return np.random.randint(0, 1000, size=size, dtype=np.int32)


def calculate_sum(data):
    return np.sum(data)


def calculate_mean(data):
    return np.mean(data)


def calculate_std(data):
    return np.std(data, ddof=0)


def do_statistical_analysis(size):
    metrics = {}
    start_total = time.perf_counter()

    start_op = time.perf_counter()
    data = create_data_structure(size)
    metrics['create'] = {
        'time': (time.perf_counter() - start_op) * 1000,
        'memory': (data.nbytes) / (1024 ** 2)
    }

    operations = {
        'sum': calculate_sum,
        'mean': calculate_mean,
        'std': calculate_std
    }

    for op_name, op_func in operations.items():
        start_op = time.perf_counter()
        result = op_func(data)
        metrics[op_name] = {
            'time': (time.perf_counter() - start_op) * 1000,
            'memory': (data.nbytes) / (1024 ** 2)
        }

    metrics['output'] = {
        'total_time': (time.perf_counter() - start_total) * 1000,
        'memory_peak': metrics['create']['memory']
    }

    result_str = ""
    for op in ['create', 'sum', 'mean', 'std']:
        result_str += f"{op.upper()} - Time: {metrics[op]['time']:.2f} ms | RAM: {metrics[op]['memory']:.2f} MB\n"

    result_str += f"TOTAL - Time: {metrics['output']['total_time']:.2f} ms | RAM Peak: {metrics['output']['memory_peak']:.2f} MB"

    return result_str


sync.do_statistical_analysis = do_statistical_analysis
